providers/implementations/signature/rsa.c

```diff
@@ -0,0 +1,922 @@
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_DIGEST_SIZE);
+    if (p != NULL && !OSSL_PARAM_get_size_t(p, &prsactx->mdsize))
+        return 0;
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_PAD_MODE);
+    if (p != NULL) {
+        int pad_mode = RSA_NO_PADDING;
+
+        if (!OSSL_PARAM_get_int(p, &pad_mode))
+            return 0;
+        switch (pad_mode) {
+        case RSA_PKCS1_OAEP_PADDING:
+            /*
+             * OAEP padding is for asymmetric cipher only so is not compatible
+             * with signature use.
+             */
+            ERR_raise(ERR_LIB_RSA,
+                      PROV_R_ILLEGAL_OR_UNSUPPORTED_PADDING_MODE);
+            return 0;
+        case RSA_PKCS1_PSS_PADDING:
+            if (prsactx->mdname[0] == '\0') {
+                OPENSSL_strlcpy(prsactx->mdname, "SHA1",
+                                sizeof(prsactx->mdname));
+                OPENSSL_strlcpy(prsactx->mdprops, "",
+                                sizeof(prsactx->mdprops));
+                EVP_MD_free(prsactx->md);
+                prsactx->md = NULL;
+            }
+            goto cont;
+        case RSA_PKCS1_PADDING:
+        case RSA_SSLV23_PADDING:
+        case RSA_NO_PADDING:
+        case RSA_X931_PADDING:
+            if (RSA_get0_pss_params(prsactx->rsa) != NULL) {
+                ERR_raise(ERR_LIB_RSA,
+                          PROV_R_ILLEGAL_OR_UNSUPPORTED_PADDING_MODE);
+                return 0;
+            }
+        cont:
+            if (!check_padding_md(prsactx->md, pad_mode))
+                return 0;
+            break;
+        default:
+            return 0;
+        }
+        prsactx->pad_mode = pad_mode;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_PSS_SALTLEN);
+    if (p != NULL) {
+        int saltlen;
+
+        if (prsactx->pad_mode != RSA_PKCS1_PSS_PADDING) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PSS_SALTLEN);
+            return 0;
+        }
+        if (!OSSL_PARAM_get_int(p, &saltlen))
+            return 0;
+
+        if (saltlen < RSA_PSS_SALTLEN_MAX) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_PSS_SALTLEN);
+            return 0;
+        }
+
+        prsactx->saltlen = saltlen;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_MGF1_DIGEST);
+    if (p != NULL) {
+        const OSSL_PARAM *propsp =
+            OSSL_PARAM_locate_const(params,
+                                    OSSL_SIGNATURE_PARAM_MGF1_PROPERTIES);
+        EVP_MD *md;
+        char *str;
+        int ret = 1;
+        char mdname[80] = "";
+        char mdprops[80] = "";
+
+        str = mdname;
+        if (!OSSL_PARAM_get_utf8_string(p, &str, sizeof(mdname)))
+            return 0;
+
+        str = mdprops;
+        if (propsp != NULL
+            && !OSSL_PARAM_get_utf8_string(propsp, &str, sizeof(mdprops)))
+            return 0;
+
+        if (prsactx->pad_mode != RSA_PKCS1_PSS_PADDING) {
```

Comment:
Is it allowable to set the digests before setting the pad mode? Or do we require the pad mode to be set first?

Reply:
Your guess is as good as mine:

https://github.com/openssl/openssl/blob/d3a27c5ee45a29edd9c0d60ad5929f67996f89fd/crypto/rsa/rsa_pmeth.c#L517-L523

Reply:
Ok...looks like we already have this restriction so, no problem.

Reply:
Note that the manual does say this for getting:

> The EVP_PKEY_CTX_get_rsa_pss_saltlen() macro gets the RSA PSS salt length
for B<ctx>. The padding mode must have been set to B<RSA_PKCS1_PSS_PADDING>.